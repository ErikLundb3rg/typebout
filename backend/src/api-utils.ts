import { NextFunction, Request, Response } from 'express';

interface BaseResponse {
  message: string | null;
  data: unknown;
  status: number;
  ok: boolean;
}

export const defaultErrorResponse = (
  overrides: Partial<Omit<BaseResponse, 'ok'>>
): BaseResponse => ({
  message: overrides.message || null,
  data: overrides.data || null,
  status: overrides.status || 400,
  ok: false
});

export const defaultHappyResponse = (
  overrides: Partial<Omit<BaseResponse, 'ok'>>
): BaseResponse => ({
  message: overrides.message || null,
  data: overrides.data || null,
  status: overrides.status || 200,
  ok: true
});

export type AsyncController<Body> = (
  req: Request<object, object, Body>,
  res: Response
) => Promise<BaseResponse>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const asyncHandler = (controller: AsyncController<any>) => {
  return async (req: Request, res: Response, next: NextFunction) => {
    try {
      const response = await controller(req, res);
      const { message, data, status, ok } = response;
      res.status(status).send({
        message,
        ok,
        data
      });
    } catch (error) {
      next(error);
    }
  };
};
