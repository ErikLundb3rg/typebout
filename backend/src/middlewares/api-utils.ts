import { NextFunction, Request, Response } from 'express'

export interface BaseResponse {
  message: string | null
  data: unknown
  status: number
  ok: boolean
}

export const defaultErrorResponse = (
  overrides: Partial<Omit<BaseResponse, 'ok'>>
): BaseResponse => ({
  message: overrides.message || null,
  data: overrides.data || null,
  status: overrides.status || 400,
  ok: false
})

export const defaultHappyResponse = (
  overrides: Partial<Omit<BaseResponse, 'ok'>>
): BaseResponse => ({
  message: overrides.message || null,
  data: overrides.data || null,
  status: overrides.status || 200,
  ok: true
})

export type AsyncController<Body = {}, QueryProps = {}> = (
  req: Request<object, object, Body, QueryProps>,
  res: Response
) => Promise<BaseResponse>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const asyncHandler = (controller: AsyncController<any, any>) => {
  return async (req: Request, res: Response, next: NextFunction) => {
    try {
      const response = await controller(req, res)
      sendBaseResponse(response, res)
    } catch (error) {
      next(error)
    }
  }
}

export const sendBaseResponse = (baseResponse: BaseResponse, res: Response) => {
  const { message, data, status, ok } = baseResponse
  res.status(status).send({
    message,
    ok,
    data
  })
}
