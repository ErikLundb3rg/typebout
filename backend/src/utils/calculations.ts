import { Users, Performances } from '@prisma/client'

export const getWPM = (letters: number, seconds: number) =>
  Math.round(letters / 6 / (seconds / 60))

export const getEnrichedPerformance = <T extends Performances>(
  performance: T
) => {
  const { correct, mistakes, completed_in_ms } = performance
  const total = correct + mistakes
  const wpm = getWPM(total, completed_in_ms / 1000)

  return {
    ...performance,
    total,
    wpm
  }
}

export const round = (value: number, precision: number) => {
  const multiplier = Math.pow(10, precision || 0)
  return Math.round(value * multiplier) / multiplier
}
