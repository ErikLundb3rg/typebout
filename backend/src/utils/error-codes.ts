export const errorCodes = {
  UNAUTHENTICATED: 401,
  BAD_REQUEST: 400,
  INTERNAL_SERVER_ERROR: 500
}
