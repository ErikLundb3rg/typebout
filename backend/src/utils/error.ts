export class BaseError extends Error {
  public status: number
  public message: string

  constructor(status: number, message?: string, stack?: string) {
    super()
    this.status = status ?? 500

    // So that `instanceof` will work with custom errors
    // https://www.dannyguo.com/blog/how-to-fix-instanceof-not-working-for-custom-errors-in-typescript
    Object.setPrototypeOf(this, BaseError.prototype)

    if (process.env.MODE === 'dev') {
      this.message = stack
        ? String(message) + '\nStack: ' + String(stack)
        : String(message)
      return
    }

    if (this.status >= 500) {
      this.message = 'Internal Server Error'
      return
    }

    this.message = String(message)
  }
}
