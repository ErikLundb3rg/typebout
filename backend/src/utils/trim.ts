// Ehhhh, sorry!
export const trimStringRecord = <T extends Record<string, string>>(
  input: T
) => {
  const trimmedObj = {} as T
  for (let k in input) {
    trimmedObj[k as keyof T] = input[k].trim() as T[keyof T]
  }
  return trimmedObj
}
