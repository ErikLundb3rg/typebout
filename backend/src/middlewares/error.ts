import { Request, Response } from 'express'
import { BaseResponse, sendBaseResponse } from './api-utils'
import { BaseError } from '../utils/error'
import { errorCodes } from '../constants/error-codes'

export const errorHandlingMiddleWare = (
  genericError: Error,
  req: Request,
  res: Response
) => {
  const response: BaseResponse = {
    data: null,
    ok: false,
    status: 500,
    message: 'Something went wrong'
  }

  const error = genericError as BaseError

  if (error instanceof SyntaxError || error instanceof URIError) {
    response.status = errorCodes.BAD_REQUEST
    response.message = 'Unprocessable request'
  } else if (error instanceof BaseError) {
    response.status = error.status
    response.message = error.message
  }

  if (process.env.MODE === 'dev') {
    console.error(error)
  }

  sendBaseResponse(response, res)
}
