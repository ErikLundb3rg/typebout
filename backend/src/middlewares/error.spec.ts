/* eslint-disable @typescript-eslint/no-explicit-any */
import { errorHandlingMiddleWare } from './error'
import { sendBaseResponse } from './api-utils'
import { BaseError } from '../utils/error'
import { errorCodes } from '../utils/error-codes'

describe('Error handling middleware', () => {
  let req: any = {}
  let res: any = {}

  beforeEach(() => {
    req = {}
    res = {
      send: jest.fn(),
      status: jest.fn()
    }
    ;(sendBaseResponse as jest.Mock) = jest.fn()
  })

  test('should return a error response on Syntax', async () => {
    errorHandlingMiddleWare(new SyntaxError(), req, res)
    const response = (sendBaseResponse as jest.Mock).mock.calls[0][0]
    console.log('response', response)
    expect(response).toMatchSnapshot()
  })
  test('should return a error response on URI error', async () => {
    errorHandlingMiddleWare(new URIError(), req, res)
    const response = (sendBaseResponse as jest.Mock).mock.calls[0][0]
    expect(response).toMatchSnapshot()
  })
  test('should replace response with BaseError', async () => {
    errorHandlingMiddleWare(
      new BaseError(errorCodes.BAD_REQUEST, 'Custom message'),
      req,
      res
    )
    const response = (sendBaseResponse as jest.Mock).mock.calls[0][0]
    expect(response).toMatchSnapshot()
  })
})
