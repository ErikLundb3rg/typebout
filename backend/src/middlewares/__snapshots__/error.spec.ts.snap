// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error handling middleware should replace response with BaseError 1`] = `
{
  "data": null,
  "message": "Custom message",
  "ok": false,
  "status": 400,
}
`;

exports[`Error handling middleware should return a error response on Syntax 1`] = `
{
  "data": null,
  "message": "Unprocessable request",
  "ok": false,
  "status": 400,
}
`;

exports[`Error handling middleware should return a error response on URI error 1`] = `
{
  "data": null,
  "message": "Unprocessable request",
  "ok": false,
  "status": 400,
}
`;
